### PR TITLE
Login and Logout 

### PR DESCRIPTION
**Pull request by magnus-thor**
### https://www.pivotaltracker.com/story/show/151510678 ###
---------------------------
* User can sign up
* User can log in
* User can log out
* Hide "log in" and "sign up" when user is logged in
* Hide "log out" when user is logged out

### CODE REVIEW

File: app/views/landing/index.html.erb [hunk 1]
```diff
@@ -5,6 +5,8 @@
   Having problems with the site? Email our system admin at unicorn_boss@rainbows.now</p>
 
 <%= link_to 'Sign Up', new_user_registration_path %>
+<%= link_to 'Log In', new_user_session_path %>
+<%= link_to "Sign out", destroy_user_session_path, :method => :delete %>
```
Comment (tochman) [inline app/views/landing/index.html.erb]: please condition this
```
<% if current_user %>
  <%= link_to "Sign out", destroy_user_session_path, :method => :delete %>
<% else %>
  <%= link_to 'Sign Up', new_user_registration_path %>
  <%= link_to 'Log In', new_user_session_path %>
<% end %>
```
Can you see why this is good?

File: features/step_definitions/user_logout_steps.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Given(/^I am on the laundry room page$/) do
+  visit '/laundry_room'
```
Comment (tochman) [inline features/step_definitions/user_logout_steps.rb]: Please use path helpers and avoid hard coding the urls

File: features/step_definitions/user_logout_steps.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+end
+
+
+Given(/^I am logged in as that user$/) do
+  user = User.last
+  login_as(user)
+end
```
Comment (tochman) [inline features/step_definitions/user_logout_steps.rb]: Saving an object in an instance variable makes it possible to access it across step definitions.

File: features/step_definitions/user_logout_steps.rb (hunk 2 above)
Comment (magnus-thor) [reply features/step_definitions/user_logout_steps.rb]: Its leftovers from earlier tests that I forgot to delete

File: features/login.feature [hunk 4]
```diff
@@ -0,0 +1,32 @@
+Feature: As a user
```
Comment (faraznaeem) [inline features/login.feature]: The Feature should state the feature, the story should be moved down see example

```
Feature: Landing page
  As a user,
  In order to search for pictures
  I need to be on the landing page
```

File: features/login.feature [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  In order to be able to do laundry
+  I would like to be able to log in to LaundryTime
+
+  Background:
+    Given There is a user with email "unicorn@test.com" and password "password" in our database
+
+  Scenario: User provides valid credentials
+    Given I navigate to the index page
+    And I click on "Log In"
```
Comment (faraznaeem) [inline features/login.feature]: Should be in the background steps
```
Given I navigate to the index page 
```

File: features/login.feature [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    And I fill in "Email" with "unicorn@test.com"
+    And I fill in "Password" with "password"
+    And I click on "Log in"
+    Then I should see the text "Signed in successfully."
+
+  Scenario: User provides wrong email
+    Given I navigate to the index page
```
Comment (faraznaeem) [inline features/login.feature]: Same as above

File: features/login.feature [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    Given I click on "Log In"
```
Comment (faraznaeem) [inline features/login.feature]: Should also be in the background steps

File: features/login.feature [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    And I fill in "Email" with "EmbarrassedUnicorn@test.com"
+    And I fill in "Password" with "password"
+    And I click on "Log in"
+    And I should see the text "Invalid Email or password."
+
+  Scenario: User provides wrong password
+    Given I navigate to the index page
+    Given I click on "Log In"
```
Comment (faraznaeem) [inline features/login.feature]: Same as other scenarios move to the background steps

File: features/login.feature [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    And I fill in "Email" with "unicorn@test.com"
+    And I fill in "Password" with "PassWord"
+    And I click on "Log in"
+    And I should see the text "Invalid Email or password."
+
```
Comment (faraznaeem) [inline features/login.feature]: Please follow this format when writing your scenarios.
```
Scenario: Buy last coffee
  Given there are 1 coffees left in the machine
  And I have deposited 1$
  When I press the coffee button
  Then I should be served a coffee
```
https://github.com/cucumber/cucumber/wiki/Feature-Introduction

File: features/user_logout.feature [hunk 10]
```diff
@@ -0,0 +1,13 @@
+Feature: As a user
```
Comment (faraznaeem) [inline features/user_logout.feature]: Please look at the above feature and make the same changes in this feature.

File: app/views/landing/index.html.erb (hunk 1 above)
Comment (Sajaas) [reply app/views/landing/index.html.erb]: Done! We actually missed your comment but decided that we wanted to hide some buttons depending on if the user is logged in or not.

**Review by faraznaeem — CHANGES_REQUESTED** [review]

**Review by faraznaeem — APPROVED** [review]

**Pull request merged**